### PR TITLE
feat(e2e_test): add new method to test with a docker image (MySQL) or Directly on Snowflake

### PR DESCRIPTION
**Pull request by f-bb-toucantoco**
Just a first case with: 
- download docker image: MySQL
- insert data in MySQL
- generate a request with translate
- execute request in MySQL

### CODE REVIEW

**Comment by f-bb-toucantoco** [comment]
[[Snowflake - VQB Quality] - POC a first case with a hard request on SQLlite (3) [3]](https://trello.com/c/w9Z9pBcL/1129-snowflake-vqb-quality-poc-a-first-case-with-a-hard-request-on-sqllite-3-3)

**Review by raphaelvignes — COMMENTED** [review]
Penser à rajouter la récupération de l'image docker du mysql dans le fichier de github actions, comme ici: 
https://github.com/ToucanToco/toucan-connectors/blob/e5b9d89b4b55d01dd8fef5bfaeab963123cdac6f/.github/workflows/ci.yml#L61

**Comment by raphaelvignes** [comment]
Faut rebase sur master pour corriger ce pb: 
TypeError: translate_pipeline() missing 1 required positional argument: 'sql_query_describer'

File: server/tests/backends/sql_translator/test_sql_translator_steps.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        if i.tags[0] == f'{image["image"]}:{image["version"]}':
+            found = True
+    if not found:
+        logging.getLogger(__name__).info(
+            f'Download docker image {image["image"]}:{image["version"]}'
+        )
+        docker_client.images.pull('mysql:5.7.21')
+
+    logging.getLogger(__name__).info(f'Start docker image {image["image"]}:{image["version"]}')
+    global docker_container
```
Comment (davinov) [inline server/tests/backends/sql_translator/test_sql_translator_steps.py:61]: Why do you need to keep it referenced in the upper scope?

File: server/tests/backends/sql_translator/test_sql_translator_steps.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    9: 'object',
+    10: 'array',
+    11: 'binary',
+    12: 'time',
+    13: 'boolean',
+}
+
+
+@pytest.fixture(scope='module', autouse=True)
+def toto():
```
Comment (davinov) [inline server/tests/backends/sql_translator/test_sql_translator_steps.py:48]:
```suggestion
def db_container():
```

File: server/tests/backends/sql_translator/test_sql_translator_steps.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    )
+    ready = False
+    while not ready:
+        time.sleep(1)
+        try:
+            if docker_container.status == 'created' and get_connection():
+                ready = True
+        except OperationalError:
+            pass
+    yield docker_container
```
Comment (davinov) [inline server/tests/backends/sql_translator/test_sql_translator_steps.py:84]: :+1:

File: server/tests/backends/sql_translator/test_sql_translator_steps.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+
+def get_engine():
+    host = '127.0.0.1'
+    user = 'ubuntu'
+    password = 'ilovetoucan'
+    port = 3306
+    database = 'mysql_db'
+    engine = create_engine(f'mysql://{user}:{password}@{host}:{port}/{database}')
+    return engine
```
Comment (davinov) [inline server/tests/backends/sql_translator/test_sql_translator_steps.py:108]: Style: could be fixture as well

File: server/tests/backends/sql_translator/test_sql_translator_steps.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+
+# Translation from Pipeline json to SQL query
+@pytest.mark.parametrize('case_id,case_spec_file_path', test_cases)
+def test_sql_translator_pipeline(case_id, case_spec_file_path):
+    spec_file = open(case_spec_file_path, 'r')
+    spec = json.loads(spec_file.read())
+    spec_file.close()
+
+    # TODO - insert data from fixture.json
```
Comment (davinov) [inline server/tests/backends/sql_translator/test_sql_translator_steps.py:169]: Done just below :D

File: server/tests/backends/sql_translator/test_sql_translator_steps.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+from weaverbird.backends.sql_translator import translate_pipeline
+from weaverbird.pipeline import Pipeline
+
+pymysql.install_as_MySQLdb()
+
+image = {'name': 'mysql_weaverbird_test', 'image': 'mysql', 'version': 'latest'}
+docker_client = docker.from_env()
+docker_container = None
+
+fixtures_dir_path = path.join(path.dirname(path.realpath(__file__)), '../fixtures_sql')
```
[the remark this answers is not in the document]
Comment (davinov) [reply server/tests/backends/sql_translator/test_sql_translator_steps.py:26]: Other proposal: it's OK ta adapt the glob to only support some fixtures and not others (for example only the ones of the steps that are implemented in SQL)
Another: add just a `"worksWithSql": true` in fixtures you want to support for now, and use this to skip the fixtures unsupported

File: server/tests/backends/sql_translator/test_sql_translator_steps.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+    query, report = translate_pipeline(
+        pipeline,
+        sql_query_retriever=sql_retrieve_city,
+        sql_query_describer=sql_query_describer,  # replace by snowflake_query_describer
+    )
+    result = execute(get_connection(), query)
+    assert query_expected == query
+
+    assert_dataframes_equals(pandas_result_expected, result)
```
Comment (davinov) [inline server/tests/backends/sql_translator/test_sql_translator_steps.py:192]: This function is great!

Maybe adding a few comments to help understanding the different steps:
- inserting data
- translating the query
- executing the query
- comparing result and expected

File: server/tests/backends/sql_translator/test_sql_translator_steps.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        cursor.execute(query)
+        # result = cursor.fetchall()
+        df = pd.DataFrame(cursor.fetchall())
+        field_names = [i[0] for i in cursor.description]
+        df.columns = field_names
+        return df
+
+
+def sql_retrieve_city(t):
+    return t
```
Comment (davinov) [inline server/tests/backends/sql_translator/test_sql_translator_steps.py:122]: What is the aim of this identify function?

File: server/weaverbird/temp.json [hunk 9]
```diff
@@ -0,0 +1,93 @@
+{
```
Comment (davinov) [inline server/weaverbird/temp.json:1]: Forgotten file?

**Review by davinov — COMMENTED** [review]
Thanks ! This is a good step in the right direction :)

I have two concerns:
- As discussed, I fear creating another set of features defeats the purpose of the fixtures that we want common to all backends. (I know you have that in mind)
- If you want to test the sql query that is returned (as string), maybe it's better to do a separate file. No need to execute any query for this, the translation is enough. It's more unit tests that "end 2 end" ones like the one we'd like to do with these fixtures.
- More about the SQL translator itself, not its tests: It's not entirely clear to me what are the  purpose of sql_query_retriever and sql_query_describer. Maybe a few docstrings or comments would help :)

**Review by raphaelvignes — APPROVED** [review]
auto-approval \0/

**Pull request merged**